src/libs/actions/Policy.js

```diff
@@ -1458,6 +1461,375 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
+                [_.keys(workspaceChatData)[0]]: {
+                    pendingAction: null,
+                },
+            },
+        },
+        ...employeeWorkspaceChat.onyxSuccessData,
+    ];
+
+    const failureData = [
+        {
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.POLICY_MEMBERS}${policyID}`,
+            value: null,
+        },
+        {
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${announceChatReportID}`,
+            value: null,
+        },
+        {
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${announceChatReportID}`,
+            value: null,
+        },
+        {
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${adminsChatReportID}`,
+            value: null,
+        },
+        {
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${adminsChatReportID}`,
+            value: null,
+        },
+        {
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${workspaceChatReportID}`,
+            value: null,
+        },
+        {
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${workspaceChatReportID}`,
+            value: null,
+        },
+    ];
+
+    // Compose the memberData object which is used to add the employee to the workspace and
+    // optimistically create the workspace chat for them.
+    const memberData = {
+        accountID: Number(employeeAccountID),
+        email: employeeEmail,
+        workspaceChatReportID: employeeWorkspaceChat.reportCreationData[employeeEmail].reportID,
+        workspaceChatCreatedReportActionID: employeeWorkspaceChat.reportCreationData[employeeEmail].reportActionID,
+    };
+
+    const oldChatReportID = iouReport.chatReportID;
+
+    // Next we need to convert the IOU report to Expense report.
+    // We need to change:
+    // - report type
+    // - change the sign of the report total
+    // - update its policyID and policyName
+    // - update the chatReportID to point to the new workspace chat
+    const expenseReport = {
+        ...iouReport,
+        chatReportID: memberData.workspaceChatReportID,
+        policyID,
+        policyName: workspaceName,
+        type: CONST.REPORT.TYPE.EXPENSE,
+        total: -iouReport.total,
+    };
+    optimisticData.push({
+        onyxMethod: Onyx.METHOD.MERGE,
+        key: `${ONYXKEYS.COLLECTION.REPORT}${iouReportID}`,
+        value: expenseReport,
+    });
+    failureData.push({
+        onyxMethod: Onyx.METHOD.MERGE,
+        key: `${ONYXKEYS.COLLECTION.REPORT}${iouReportID}`,
+        value: iouReport,
+    });
+
+    // The expense report transactions need to have the amount reversed to negative values
+    const reportTransactions = TransactionUtils.getAllReportTransactions(iouReportID);
```

Comment:
Would that be beneficial here? we do not need to have another connection to Transactions collection in Policy.js but just get all the transactions linked to the report, right?

Reply:
There are several benefits to adding a new connection in this file:

- The local structure can be optimized for reportID lookup (ie. store them in a map like `{<reportID>: transactions}` rather than having two `filters()` run every time in `getAllReportTransactions()`
- The action files remain decoupled (something we do a terrible job of)
- Decoupled actions keep regressions to a minimum